File: c++/quick_sort.cpp
/*
Quick sort is a highly efficient sorting algorithm and is based on partitioning of array of data into smaller arrays. A large array is
partitioned into two arrays, one of which holds values smaller than specified value, say pivot, based on which the partition is made,
and another array holds values greater than the pivot value.

Quick sort partitions an array and then makes two recursive calls to sort the two resulting sub-arrays. This algorithm is quite efficient
for large-sized data sets as its average and worst-case complexity are O(n^2).


**** quick sort ****
*/
#include <stdio.h>
#include<stdlib.h>
#include <iostream>

using namespace std;

void swap(int *x,int *y) {
    int temp = *x;
    *x = *y;
    *y = temp;
}

int partition(int A[],int l,int h) {
    int pivot = A[l];
    int i = l,j = h;

    do {
        do {
            i++;
        }
        while(A[i] <= pivot);

        do{
            j--;
        } while(A[j] > pivot);

        if (i < j)
            swap(&A[i],&A[j]);
    } while(i < j);

    swap(&A[l],&A[j]);
    return j;
}

void quickSort(int A[],int l,int h) {
    int j;

    if (l < h) {
        j = partition(A,l,h);
        quickSort(A,l,j);
        quickSort(A,j + 1,h);
    }
}

int main() {
    cout << "**** quick sort ****\n";
    int A[] = {9,14,77,0,-4,25,30,59,2,45,12,23},numElem = 12,i;

     cout << "before sort: " << endl;
     for (i = 0; i < numElem; i++)
        cout << A[i] << " ";

    cout << "\nafter sort: \n";
    quickSort(A,0,numElem);
    for(i = 0; i < numElem; i++)
        printf("%d ",A[i]);

    printf("\n");
    return 0;
}

File: java/QuickSort.java
/*





**** quick sort ****

 */
import java.util.Arrays;

public class QuickSort {
  // get pivot point
  static int partition(int[] array,int start,int end) {
    int pivot = end;
    int i = start - 1;
    for (int j = start; j <= end; j++) {
      if (array[j] <= array[pivot]) {
        i++;
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
      }
    }

    return i;
  }

  public static void beginRecurse(int[] array,int start,int end) {
    if (start < end) {
      int pivot = partition(array,start,end);
      beginRecurse(array,start,pivot - 1);
      beginRecurse(array,pivot + 1,end);
    }
  }

  public static void printArray(int[] array) {
      for (int i = 0; i < array.length; i++) {
          System.out.print(array[i] + " ");
      }
  }

  public static void main(String[] args) {
      System.out.println("**** quick sort ****");
      QuickSort qs = new QuickSort();
      int arr[] = {4,25,0,37,44,-11,19,89,7,9};

      System.out.println("before sort: " + Arrays.toString(arr));
      qs.beginRecurse(arr,0,arr.length - 1);
      System.out.println("after sort: " + Arrays.toString(arr));
  }
}
